Indicators/SortinoRatio.cs

```diff
@@ -0,0 +1,121 @@
+    {
+        private readonly int _period;
+
+        /// <summary>
+        /// RateOfChange indicator for calculating sma for the sortino ratio
+        /// </summary>
+        private readonly RateOfChange _rocForSMA;
+
+        /// <summary>
+        /// RateOfChange indicator for calculating downside deviation for the sortino ratio
+        /// </summary>
+        private readonly RateOfChange _rocForSTD;
+
+        /// <summary>
+        /// Indicator to store the calculation of the sortino ratio
+        /// </summary>
+        private readonly CompositeIndicator _sortinoRatio;
+
+        /// <summary>
+        /// Required period, in data points, for the indicator to be ready and fully initialized.
+        /// </summary>
+        public int WarmUpPeriod { get; }
+
+        /// <summary>
+        /// Returns whether the indicator is properly initialized with data
+        /// </summary>
+        public override bool IsReady => _sortinoRatio.Samples > _period;
+
+        /// <summary>
+        /// Creates a new Sortino Ratio indicator using the specified periods
+        /// </summary>
+        /// <param name="period">Period of historical observation for sortino ratio calculation</param>
+        /// <param name="riskFreeRate">Risk-free rate for sortino ratio calculation</param>
+        public SortinoRatio(int period, decimal riskFreeRate = 0.0m)
+            : this($"SORTINO({period},{riskFreeRate})", period, riskFreeRate)
+        {
+        }
+
+        /// <summary>
+        /// Creates a new Sortino Ratio indicator using the specified periods
+        /// </summary>
+        /// <param name="name">The name of this indicator</param>
+        /// <param name="period">Period of historical observation for sortino ratio calculation</param>
+        /// <param name="riskFreeRate">Risk-free rate for sortino ratio calculation</param>
+        public SortinoRatio(string name, int period, decimal riskFreeRate = 0.0m)
+            : base(name)
+        {
+            _period = period;
+
+            // calculate sortino ratio using indicators
+            _rocForSMA = new RateOfChange(1);
+            _rocForSTD = new RateOfChange(1);
+            _rocForSTD.Updated += (sender, args) =>
+            {                
+                //ensure we are only using the downside deviation for calculation
+                var _roc = sender.ConvertInvariant<RateOfChange>();
+                if (_roc.Current.Value > riskFreeRate)
+                    _roc.Current.Value = riskFreeRate;
+            };
+            var std = new StandardDeviation(period).Of(_rocForSTD);
+            var sma = _rocForSMA.SMA(period);
+            _sortinoRatio = sma.Minus(riskFreeRate).Over(std);
+
+            // define warmup value; 
+            // _roc is the base of our indicator chain + period of STD and SMA
+            WarmUpPeriod = _rocForSMA.WarmUpPeriod + _period;
```

Comment:
Cool!
- The implementation is **very** similar to Sharpe Ratio one, seems they could share their implementation with some inheritance maybe, what would change is the `StandardDeviation` vs `StandardDownsideDeviation`?
- `std/sma/_roc/_rocForSTD/_rocForSMA` following Lean coding style ideally these should be expanded to their full version, and maybe instead of their name being they type it could reflect their purpose/nature?
- Think the `Standard Downside Deviation` should be it's own indicator with his tests? Not sure the currently implementation is right, based on https://www.investopedia.com/terms/d/downside-deviation.asp should only take negative returns after subtracting the free rate?
- `_rocForSTD` ideally we shouldn't be changing the indicators value after it's being calculated, seems like this should live inside the indicator or in the consumer of it's value

Reply:
I was about to suggest only one `RateOfChange` to simplify the code a little bit, and remove the recalculation of `_rocForStd`, but the idea of implementing `Standard Downside Deviation` is very good since it might be used in other applications.

Reply:
Thanks @AlexCatarino, I’ll have some updated changes pushed soon!